### PR TITLE
Add distributed Scatter/Gather 

### PR DESCRIPTION
**Pull request by szaman19**
- Enables distconv implementations of Scatter and Gather layers
- Implements NVSHMEM based RMA kernels for scatter/gather on DiHydrogen tensors
- Adds example applications in `applications/graph/DistConvGNN/synthetic` for benchmarking distributed Scatter, Gather, and GCN
- Adds unit tests

To do:
- [x] Fix error in distconv identity layer causing mismatched mini-batch dimension

### CODE REVIEW

**Comment by szaman19** [comment]
Tagging as this affected by #2145

File: include/lbann/layers/transform/distconv/distconv_gather.hpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+#ifndef LBANN_LAYERS_TRANSFORM_DISTCONV_GATHER
+#define LBANN_LAYERS_TRANSFORM_DISTCONV_GATHER
+#include "lbann/utils/distconv.hpp"
+#include "distconv/base.hpp"
+#include "distconv/tensor/tensor.hpp"
+#include "distconv/tensor/tensor_mpi.hpp"
+#include "lbann/layers/transform/distconv/distconv_nvshmem_vector_addressing.hpp"
+
+#ifdef LBANN_HAS_DISTCONV
```
Comment (benson31) [inline include/lbann/layers/transform/distconv/distconv_gather.hpp:35]: Should this have a guard for NVSHMEM, too? I.e., will `tensor::ScatterNVSHMEM` be defined-enough if we don't have NVSHMEM and this file gets `#include`-ed somewhere?

File: include/lbann/layers/transform/distconv/distconv_scatter.hpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+#ifndef LBANN_LAYERS_TRANSFORM_DISTCONV_SCATTER
+#define LBANN_LAYERS_TRANSFORM_DISTCONV_SCATTER
+#include "lbann/utils/distconv.hpp"
+#include "distconv/base.hpp"
+#include "distconv/tensor/tensor.hpp"
+#include "distconv/tensor/tensor_mpi.hpp"
+#include "lbann/layers/transform/distconv/distconv_nvshmem_vector_addressing.hpp"
+
+#ifdef LBANN_HAS_DISTCONV
```
Comment (benson31) [inline include/lbann/layers/transform/distconv/distconv_scatter.hpp:35]: Same question as `distconv_gather.hpp`

**Review by bvanessen — APPROVED** [review]

File: include/lbann/layers/transform/scatter.hpp [hunk 3]
```diff
@@ -128,31 +162,81 @@ El::Device scatter_layer<TensorDataType,Layout,Device>::get_device_allocation()
...
   const auto& input1_dims = this->get_input_dims(1);
+  const auto& output_dims = this->get_output_dims();
+
+  auto dims_to_str = [] (const std::vector<int>& dims) -> std::string {
+    std::ostringstream ss;
+    for (size_t i=0; i<dims.size(); ++i) {
+      ss << (i>0 ? "x" : "") << dims[i];
+    }
+    return ss.str();
+  };
```
Comment (benson31) [inline include/lbann/layers/transform/scatter.hpp:177]: Don't we have a function for this? If no, we should have one at the library utils level...

File: src/layers/transform/distconv/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  distconv_gather.cpp
+  distconv_scatter.cpp
+)
+set_full_path(THIS_DIR_CU_SOURCES
+  distconv_nvshmem_vector_addressing.cu
+)
+endif()
+# Propagate the files up the tree
+set(SOURCES "${SOURCES}" "${THIS_DIR_SOURCES}" PARENT_SCOPE)
+set(GPU_SOURCES "${GPU_SOURCES}" "${THIS_DIR_CU_SOURCES}" PARENT_SCOPE)
```
Comment (benson31) [inline src/layers/transform/distconv/CMakeLists.txt:14]:
```suggestion
set(GPU_SOURCES "${GPU_SOURCES}" "${THIS_DIR_CU_SOURCES}" PARENT_SCOPE)

```

File: src/layers/transform/distconv/distconv_gather.cpp [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    const auto& values_shape = values.get_local_shape();    // Should be {1, F, N, B}
+    const auto& indices_shape = indices.get_local_shape();  // Should be {1, 1, E, B}
+    const auto& output_shape = output.get_local_shape();    // Should be {1, F, E, B}
+
+    // Debug prints -- delete before PR
+    #if 0
+       util::MPIRootPrintStreamInfo() << "Values Dims: " << values_shape
+      << "\tIndices Dims: " << indices_shape << "\tOutput Dims: "<< output_shape; 
+    #endif
+
```
Comment (benson31) [inline src/layers/transform/distconv/distconv_gather.cpp:64]:
```suggestion
```

File: src/layers/transform/distconv/distconv_gather.cpp [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+  template int Gather <Backend, T>::backward<tensor::CUDAAllocator>(                      \
+    const tensor::Tensor<T, tensor::LocaleMPI, tensor::CUDAAllocator> &output_grad,       \
+    const tensor::Tensor<T, tensor::LocaleMPI, tensor::CUDAAllocator> &indices,           \
+    tensor::Tensor<T, tensor::LocaleMPI, tensor::CUDAAllocator> &values_grad,             \
+    tensor::Tensor<T, tensor::LocaleMPI, tensor::CUDAAllocator> &indices_grad);
+
+ETI(float, cudnn::BackendCUDNN)
+ETI(double, cudnn::BackendCUDNN)
+#undef ETI
+} // namespace distconv
```
Comment (benson31) [inline src/layers/transform/distconv/distconv_gather.cpp:206]:
```suggestion
} // namespace distconv

```

File: src/layers/transform/CMakeLists.txt [hunk 7]
```diff
@@ -96,6 +96,11 @@ endif ()
 # Add the subdirectories
 add_subdirectory(cereal_registration)
 
+if (LBANN_HAS_DISTCONV)
```
Comment (benson31) [inline src/layers/transform/CMakeLists.txt:99]: Did we need `NVSHMEM` protection here, too?

**Review by benson31 — COMMENTED** [review]
I would like to verify this builds `+distconv~nvshmem`.

File: include/lbann/layers/transform/scatter.hpp (hunk 3 above)
Comment (szaman19) [reply include/lbann/layers/transform/scatter.hpp:192]: Not that I know of, but this pattern does get used quite a bit. Ex:
https://github.com/LLNL/lbann/blob/4acbbe5fcb0b48af35a1afa48e628d789ecaf0f1/include/lbann/layers/transform/concatenate.hpp#L226

https://github.com/LLNL/lbann/blob/4acbbe5fcb0b48af35a1afa48e628d789ecaf0f1/include/lbann/layers/transform/sum.hpp#L116

Comment (benson31) [reply include/lbann/layers/transform/scatter.hpp:192]: That's fantastic. Not your problem, but we should probably functionify that...

File: src/layers/transform/CMakeLists.txt (hunk 7 above)
Comment (szaman19) [reply src/layers/transform/CMakeLists.txt:99]: Yes, for now. Possibly not true in the future

**Pull request merged**